File: server/src/test/java/com/fluffy/comment/application/ExamCommentServiceTest.java
package com.fluffy.comment.application;


import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;

import com.fluffy.auth.domain.Member;
import com.fluffy.auth.domain.MemberRepository;
import com.fluffy.comment.application.dto.CreateExamCommentRequest;
import com.fluffy.comment.application.dto.CreateExamCommentResponse;
import com.fluffy.comment.application.dto.DeleteExamCommentRequest;
import com.fluffy.comment.domain.ExamComment;
import com.fluffy.comment.domain.ExamCommentRepository;
import com.fluffy.exam.domain.Exam;
import com.fluffy.exam.domain.ExamRepository;
import com.fluffy.exam.domain.Question;
import com.fluffy.global.exception.ForbiddenException;
import com.fluffy.support.AbstractIntegrationTest;
import com.fluffy.support.data.MemberTestData;
import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

class ExamCommentServiceTest extends AbstractIntegrationTest {

    @Autowired
    private ExamCommentService examCommentService;

    @Autowired
    private MemberRepository memberRepository;

    @Autowired
    private ExamRepository examRepository;

    @Autowired
    private ExamCommentRepository examCommentRepository;

    @Test
    @DisplayName("댓글을 작성할 수 있다.")
    void createRootComment() {
        // given
        Member member = createMember();
        Exam exam = createExam(member, true);

        // when
        CreateExamCommentRequest request = new CreateExamCommentRequest("댓글", exam.getId(), member.getId(), null);
        CreateExamCommentResponse response = examCommentService.createComment(request);

        // then
        assertAll(
                () -> assertThat(response.id()).isNotNull(),
                () -> assertThat(response.content()).isEqualTo("댓글"),
                () -> assertThat(response.examId()).isEqualTo(exam.getId()),
                () -> assertThat(response.parentCommentId()).isNull(),
                () -> assertThat(response.author().id()).isEqualTo(member.getId()),
                () -> assertThat(response.author().name()).isEqualTo(member.getName()),
                () -> assertThat(response.author().avatarUrl()).isEqualTo(member.getAvatarUrl()),
                () -> assertThat(examCommentRepository.findAll()).hasSize(1)
        );
    }

    @Test
    @DisplayName("출제되지 않은 시험에는 댓글을 작성할 수 없다.")
    void createCommentWithNotPublishedExam() {
        // given
        Member member = createMember();
        Exam exam = createExam(member, false);

        // when, then
        CreateExamCommentRequest request = new CreateExamCommentRequest("댓글", exam.getId(), member.getId(), null);

        assertThatThrownBy(() -> examCommentService.createComment(request))
                .isInstanceOf(ForbiddenException.class)
                .hasMessage("출제되지 않은 시험에는 댓글을 작성할 수 없습니다.");

    }

    @Test
    @DisplayName("답글을 작성할 수 있다.")
    void createReplyComment() {
        // given
        Member member = createMember();
        Exam exam = createExam(member, true);
        ExamComment root = examCommentRepository.save(ExamComment.create("댓글", exam.getId(), member.getId()));

        // when
        CreateExamCommentRequest request = new CreateExamCommentRequest(
                "답글",
                exam.getId(),
                member.getId(),
                root.getId()
        );
        CreateExamCommentResponse response = examCommentService.createComment(request);

        // then
        assertAll(
                () -> assertThat(response.parentCommentId()).isEqualTo(root.getId()),
                () -> assertThat(examCommentRepository.findAll()).hasSize(2)
        );
    }

    @Test
    @DisplayName("댓글을 삭제할 수 있다.")
    void deleteComment() {
        // given
        Member member = createMember();
        Exam exam = createExam(member, true);
        ExamComment examComment = examCommentRepository.save(ExamComment.create("댓글", exam.getId(), member.getId()));

        // when
        DeleteExamCommentRequest deleteExamCommentRequest = new DeleteExamCommentRequest(
                examComment.getId(),
                member.getId()
        );
        Long deletedCommentId = examCommentService.deleteComment(deleteExamCommentRequest);

        // then
        ExamComment deletedExamComment = examCommentRepository.findByIdOrThrow(examComment.getId());
        assertAll(
                () -> assertThat(deletedCommentId).isEqualTo(examComment.getId()),
                () -> assertThat(deletedExamComment.isDeleted()).isTrue()
        );
    }

    @Test
    @DisplayName("댓글은 해당 댓글의 작성자만 삭제할 수 있다.")
    void deleteCommentWithNotWriter() {
        // given
        Member member = createMember();
        Exam exam = createExam(member, true);
        ExamComment examComment = examCommentRepository.save(ExamComment.create("댓글", exam.getId(), member.getId()));

        // when, then
        Long anotherMemberId = member.getId() + 1;
        DeleteExamCommentRequest request = new DeleteExamCommentRequest(examComment.getId(), anotherMemberId);

        assertThatThrownBy(() -> examCommentService.deleteComment(request))
                .isInstanceOf(ForbiddenException.class)
                .hasMessage("댓글 작성자만 삭제할 수 있습니다.");
    }

    private Member createMember() {
        Member member = MemberTestData.defaultMember().build();

        return memberRepository.save(member);
    }

    private Exam createExam(Member member, boolean isPublished) {
        Exam exam = Exam.create("시험", member.getId());
        exam.updateQuestions(List.of(Question.shortAnswer("질문", "설명", "정답")));

        if (isPublished) {
            exam.publish();
        }

        return examRepository.save(exam);
    }
}

File: server/src/test/kotlin/com/fluffy/comment/application/ExamCommentServiceTest.kt
package com.fluffy.comment.application

import com.fluffy.auth.domain.MemberRepository
import com.fluffy.auth.domain.findByIdOrThrow
import com.fluffy.auth.fixture.MemberFixture
import com.fluffy.comment.application.request.CreateExamCommentRequest
import com.fluffy.comment.application.request.DeleteExamCommentRequest
import com.fluffy.comment.application.response.CreateExamCommentResponse
import com.fluffy.comment.domain.ExamCommentRepository
import com.fluffy.comment.domain.findByIdOrThrow
import com.fluffy.comment.fixture.ExamCommentFixture
import com.fluffy.exam.domain.Exam
import com.fluffy.exam.domain.ExamRepository
import com.fluffy.exam.domain.findByIdOrThrow
import com.fluffy.global.exception.BadRequestException
import com.fluffy.global.exception.ForbiddenException
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk

class ExamCommentServiceTest : BehaviorSpec({

    val memberRepository = mockk<MemberRepository>()
    val examRepository = mockk<ExamRepository>()
    val examCommentRepository = mockk<ExamCommentRepository>()
    val examCommentService = ExamCommentService(memberRepository, examRepository, examCommentRepository)

    Given("createComment") {
        val request = mockk<CreateExamCommentRequest>()
        val exam = mockk<Exam>()
        val memberId = 1L
        val examId = 1L
        val member = MemberFixture.create(id = memberId)

        every { request.content } returns "댓글입니다."
        every { request.memberId } returns memberId
        every { request.examId } returns examId
        every { memberRepository.findByIdOrThrow(any()) } returns member
        every { examRepository.findByIdOrThrow(any()) } returns exam

        When("시험이 출제되지 않은 경우") {
            every { exam.isNotPublished() } returns true

            Then("예외를 발생시킨다.") {
                shouldThrow<BadRequestException> { examCommentService.createComment(request) }
                    .message shouldBe "출제되지 않은 시험에는 댓글을 작성할 수 없습니다."
            }
        }

        When("시험이 출제된 경우") {
            every { exam.isNotPublished() } returns false

            And("부모 댓글 식별자가 없는 경우") {
                every { request.parentCommentId } returns null

                val rootComment = ExamCommentFixture.create(
                    id = 1L,
                    content = request.content,
                    memberId = memberId,
                    examId = examId
                )

                every { examCommentRepository.save(any()) } returns rootComment

                val actual = examCommentService.createComment(request)
                val expected = CreateExamCommentResponse.of(rootComment, member)

                Then("루트 댓글을 생성하고, 댓글 정보를 반환한다.") {
                    actual shouldBe expected
                }
            }

            And("부모 댓글 식별자가 있는 경우") {
                val parentCommentId = 2L

                val parentComment = ExamCommentFixture.create(id = parentCommentId, examId = examId)
                val replyComment = ExamCommentFixture.create(
                    id = 3L,
                    content = request.content,
                    memberId = memberId,
                    parentCommentId = parentCommentId
                )

                every { request.parentCommentId } returns parentCommentId
                every { examCommentRepository.findByIdOrThrow(any()) } returns parentComment
                every { examCommentRepository.save(any()) } returns replyComment

                val actual = examCommentService.createComment(request)
                val expected = CreateExamCommentResponse.of(replyComment, member)

                Then("댓글에 대한 답글을 생성하고, 댓글 정보를 반환한다.") {
                    actual shouldBe expected
                }
            }
        }
    }

    Given("deleteComment") {

        When("댓글 작성자가 댓글을 삭제하는 경우") {
            val memberId = 1L
            val commentId = 1L
            val examComment = ExamCommentFixture.create(id = commentId, memberId = memberId)
            val request = DeleteExamCommentRequest(commentId, memberId)

            every { examCommentRepository.findByIdOrThrow(any()) } returns examComment

            val actual = examCommentService.deleteComment(request)

            Then("댓글을 삭제 처리하고, 댓글 식별자를 반환한다.") {
                actual shouldBe commentId
            }
        }

        When("댓글 작성자가 아닌 사용자가 댓글을 삭제하는 경우") {
            val memberId = 1L
            val commentId = 1L
            val examComment = ExamCommentFixture.create(id = commentId, memberId = 2L)
            val request = DeleteExamCommentRequest(commentId, memberId)

            every { examCommentRepository.findByIdOrThrow(any()) } returns examComment

            Then("예외를 발생시킨다.") {
                shouldThrow<ForbiddenException> { examCommentService.deleteComment(request) }
                    .message shouldBe "댓글 작성자만 삭제할 수 있습니다."
            }
        }
    }
})
